File: apps/clinical-narrative/lib/core/models/template_models.dart
library;

class TemplateRecord {
  TemplateRecord({
    required this.id,
    required this.templateGroupId,
    required this.name,
    required this.documentType,
    required this.version,
    required this.status,
    required this.body,
    required this.placeholders,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String templateGroupId;
  final String name;
  final String documentType;
  final String version;
  final String status;
  final String body;
  final List<String> placeholders;
  final String createdAt;
  final String updatedAt;

  factory TemplateRecord.fromJson(Map<String, dynamic> json) {
    return TemplateRecord(
      id: json['_id']?.toString() ?? '',
      templateGroupId: json['templateGroupId']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      documentType: json['documentType']?.toString() ?? '',
      version: json['version']?.toString() ?? '',
      status: json['status']?.toString() ?? '',
      body: json['body']?.toString() ?? '',
      placeholders: (json['placeholders'] as List?)
              ?.map((item) => item.toString())
              .toList() ??
          const [],
      createdAt: json['createdAt']?.toString() ?? '',
      updatedAt: json['updatedAt']?.toString() ?? '',
    );
  }
}

class TemplatePreview {
  TemplatePreview({
    required this.html,
    required this.title,
    required this.body,
    required this.missingFields,
    required this.metadata,
  });

  final String html;
  final String title;
  final String body;
  final List<String> missingFields;
  final Map<String, dynamic> metadata;

  factory TemplatePreview.fromJson(Map<String, dynamic> json) {
    return TemplatePreview(
      html: json['html']?.toString() ?? '',
      title: json['title']?.toString() ?? '',
      body: json['body']?.toString() ?? '',
      missingFields: (json['missingFields'] as List?)
              ?.map((item) => item.toString())
              .toList() ??
          const [],
      metadata: json['metadata'] is Map<String, dynamic>
          ? Map<String, dynamic>.from(json['metadata'] as Map)
          : const {},
    );
  }
}

File: services/survey-backend/app/domain/models/template_models.py
from datetime import datetime
from typing import Optional, List, Dict, Any

from pydantic import BaseModel, Field, ConfigDict


class TemplateRecord(BaseModel):
    id: Optional[str] = Field(default=None, alias="_id")
    template_group_id: Optional[str] = Field(default=None, alias="templateGroupId")
    name: str
    document_type: str = Field(..., alias="documentType")
    version: str = Field(default="1.0.0")
    status: str = Field(default="draft")
    body: str
    placeholders: List[str] = Field(default_factory=list)
    conditions: Optional[List[Dict[str, Any]]] = None
    created_at: datetime = Field(default_factory=datetime.utcnow, alias="createdAt")
    updated_at: datetime = Field(default_factory=datetime.utcnow, alias="updatedAt")
    approved_at: Optional[datetime] = Field(default=None, alias="approvedAt")
    created_by: Optional[str] = Field(default=None, alias="createdBy")
    updated_by: Optional[str] = Field(default=None, alias="updatedBy")
    approved_by: Optional[str] = Field(default=None, alias="approvedBy")

    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class TemplateCreateRequest(BaseModel):
    name: str
    document_type: str = Field(..., alias="documentType")
    body: str
    placeholders: List[str] = Field(default_factory=list)
    conditions: Optional[List[Dict[str, Any]]] = None

    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class TemplateUpdateRequest(BaseModel):
    name: Optional[str] = None
    document_type: Optional[str] = Field(default=None, alias="documentType")
    body: str
    placeholders: List[str] = Field(default_factory=list)
    conditions: Optional[List[Dict[str, Any]]] = None

    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class TemplatePreviewRequest(BaseModel):
    sample_data: Dict[str, Any] = Field(default_factory=dict, alias="sampleData")

    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class TemplatePreviewResponse(BaseModel):
    html: str
    title: str
    body: str
    missing_fields: List[str] = Field(default_factory=list, alias="missingFields")
    metadata: Dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class TemplateAuditRecord(BaseModel):
    id: Optional[str] = Field(default=None, alias="_id")
    template_id: Optional[str] = Field(default=None, alias="templateId")
    template_group_id: Optional[str] = Field(default=None, alias="templateGroupId")
    version: Optional[str] = None
    action: str
    status: Optional[str] = None
    actor: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow, alias="createdAt")
    metadata: Dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="forbid", populate_by_name=True)
